### PR TITLE
feat(motion): Add Stagger choreography component

### PR DESCRIPTION
**Pull request by robertpenner**
This introduces the new `Stagger` choreography component to the `react-motion-components-preview` package. 
 enabling staggered entrance and exit animations for child elements. 

In our atomic motion system, `Stagger` is an "organism" that orchestrates motion molecules like `Fade`, `Slide` for groups of content.

## New API
- The `Stagger` component manages the entrance and exit of children with configurable timing, direction, and hide/delay modes. 
  - With presence motion components like Fade and Slide, Stagger triggers their animations in sequence.
  - With plain HTML elements, Stagger shows/hides elements in order to simulate a staggered effect.
  - Stagger auto-detects optimal animation strategies based on its children, with optional overrides.
- A `Stagger` is a presence, so it includes `<Stagger.In>` and `<Stagger.Out>` components for one-way animations. 

## Documentation

- Stories start with minimalist feature-focused examples, progressing to more complex and holistic scenarios.

## Testing

- Added unit tests for the `Stagger` component, covering prop passing, mode detection, edge cases, and style preservation.

### CODE REVIEW

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/index.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+export { useStaggerItemsVisibility } from './useStaggerItemsVisibility';
+export type { UseStaggerItemsVisibilityParams } from './useStaggerItemsVisibility';
+
+// Export utilities for advanced usage
+export {
+  DEFAULT_ITEM_DELAY,
+  DEFAULT_ITEM_DURATION,
+  toElementArray,
+  isFragment,
+  childrenOrFragmentToArray,
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/index.ts]: Let's avoid exporting these for now

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/react-children-utils.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react';
+
+/**
+ * Flattens ReactNode (including Fragments) to an array of valid ReactElements,
+ * filtering out strings, numbers, null, etc.
+ */
+export const toElementArray = (children: React.ReactNode): React.ReactElement[] =>
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/react-children-utils.ts]: I don't see anything else being used yet, except `toElementArray`. Let's keep only it.

Also suggest to rewrite it to emit a warning on invalid `children`. Something like:

```
export const toElementArray = (children: React.ReactNode): React.ReactElement[] => {
  const result: React.ReactElement[] = [];

  React.Children.forEach(children, child => {
    if (React.isValidElement(child)) {
      result.push(child);
    }

    if (process.env.NODE_ENV !== 'production') {
      // eslint-disable-next-line no-console
      console.error(['toElementArray: Ignoring non-ReactElement child:', child]);
    }
  });

  return result;
};
```

Otherwise people won't know that something goes wrong

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      if (itemCount > prev.length) {
+        // Add new items in their target state
+        const targetState = direction === 'enter';
+        return [...prev, ...Array(itemCount - prev.length).fill(targetState)];
+      } else {
+        // Remove items from the end
+        return prev.slice(0, itemCount);
+      }
+    });
+  }, [itemCount, direction]);
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: IMO `itemCount` is not a reliable approach as you don't know _what's_ actually gets added/removed. This might be a problem when are being added & removed simultaneously. `getChildMapping` might be more robust approach

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        itemDelay,
+        itemDuration,
+        direction,
+        reversed,
+      });
+
+      setItemsVisibility(result.itemsVisibility);
+
+      if (elapsed < result.totalDuration) {
+        frameRef.current = requestAnimationFrame(tick as () => void);
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: Let's fix types on `useAnimationFrame()` instead:

```
export function useAnimationFrame(): readonly [(fn: FrameRequestCallback) => number, () => void] {
  const { targetDocument } = useFluent();
  const win = targetDocument?.defaultView;

  const setAnimationFrame = win ? win.requestAnimationFrame : setAnimationFrameNoop;
  const clearAnimationFrame = win ? win.cancelAnimationFrame : cancelAnimationFrameNoop;

  return useBrowserTimer(setAnimationFrame, clearAnimationFrame) as [(fn: FrameRequestCallback) => number, () => void];
}
```

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/stagger-calculations.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * Uses the formula:
+ *   max(0, delay * (count - 1) + itemDuration)
+ *
+ * @param params.count        Total number of items to stagger
+ * @param params.delay        Milliseconds between the start of each item
+ * @param params.itemDuration Milliseconds each item's animation lasts (default 0)
+ * @returns                   Total duration in milliseconds (never negative)
+ */
+export function getStaggerTotalDuration({
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/stagger-calculations.ts]: image

JSDoc there and in other places is invalid

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        cancelAnimationFrame();
+      }
+    };
+  }, [
+    animationKey,
+    itemDelay,
+    itemDuration,
+    direction,
+    reversed,
+    onMotionFinish,
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: There or in another place you probably want to wrap `onMotionFinish` with `useEventCallback`, otherwise an effect will be triggered on a reference change

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * Checks if a React element is a motion component created by createMotionComponent.
+ * Motion components have the function name "Atom".
+ *
+ * **Note:** This is a heuristic detection. Motion components may or may not support
+ * specific props like `delay` or `visible` depending on their implementation.
+ *
+ * @internal - Exported for testing purposes
+ */
+export function isMotionComponent(element: React.ReactElement): boolean {
+  return typeof element.type === 'function' && element.type.name === 'Atom';
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: This is not reliable as `.name` will be changed in production build by a minifier

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * - Elements with explicit delay and exitDelay props already set
+ *
+ * **When to override:** If auto-detection is incorrect, use explicit `delayMode` prop on Stagger.
+ * For example, custom motion components that don't support both props should use `delayMode="timing"`.
+ *
+ * @internal - Exported for testing purposes
+ */
+export function acceptsDelayProps(element: React.ReactElement): boolean {
+  return (
+    isPresenceComponent(element) || isMotionComponent(element) || hasExplicitProps(element, ['delay', 'exitDelay'])
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: `hasExplicitProps()` is a quite fragile thing. I understand that you want to have for custom components, but a better way would be to introduce a component factory to enforce the contract:

```tsx
// naming is up to you
function createStaggerChild() {}

const CustomChild = createStaggerChild(({ exitDelay, delay }) => <MyComponent delay={delay} exitDelay={exitDelay} />)
```

File: packages/react-components/react-motion-components-preview/stories/src/choreography/Stagger/StaggerHideMode.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+          <div className={classes.description}>
+            Keeps elements mounted and toggles CSS visibility to preserve layout.
+          </div>
+        </div>
+
+        <div className={classes.section}>
+          <div className={classes.sectionTitle}>unmount</div>
+          <div className={classes.items}>
+            <div className={classes.badge}>Override</div>
+            <Stagger visible={visible} hideMode="unmount" itemDelay={itemDelay} reversed={reversed}>
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/stories/src/choreography/Stagger/StaggerHideMode.stories.tsx:146]: This variant seems to be broken on initial mount:


https://github.com/user-attachments/assets/f1298863-5836-41ff-98fe-cbc33c9fb989

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/react-children-utils.ts (hunk 2 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/react-children-utils.ts]: Done.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx (hunk 4 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: Done.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/stagger-calculations.ts (hunk 5 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/stagger-calculations.ts]: Done.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts (hunk 7 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: Done.
Good catch, I was going to migrate this to use the MOTION_DEFINITION symbol after adding the PRESENCE_DEFINITION symbol.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/index.ts (hunk 1 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/index.ts]: Done.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx (hunk 3 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: Done.
I added `getStaggerChildMapping` which is based on `getChildMapping` from `react-motion`. We can consider unifying them by making it generic, but I don't want to get into changing the API for `react-motion` in this PR.

The refactoring to `getStaggerChildMapping` allowed me to delete a number of util functions.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts (hunk 8 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: I'm not sure how `hasExplicitProps()` is fragile; I am covering expected usage with unit tests.
In any case, this auto-detection is for convenience, but consumers can always set `delayMode` to `delayProp` explicitly.

File: packages/react-components/react-motion-components-preview/stories/src/choreography/Stagger/StaggerHideMode.stories.tsx (hunk 9 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/stories/src/choreography/Stagger/StaggerHideMode.stories.tsx:146]: Done.
I normalized the hide modes to use the final state for their first render (to match what `visibleProp` hide mode was already doing).

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx (hunk 6 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: @layershifter Done in 9140c5b3d6d4e08b96770bc0c00654166136a189

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+  itemDuration = DEFAULT_ITEM_DURATION,
+  direction,
+  reversed = false,
+  onMotionFinish,
+  hideMode = 'visibleProp',
+}: UseStaggerItemsVisibilityParams): { itemsVisibility: Record<string, boolean> } {
+  const [requestAnimationFrame, cancelAnimationFrame] = useAnimationFrame();
+
+  // Stabilize the callback reference to avoid re-triggering effects on every render
+  const stableOnMotionFinish = useEventCallback(
```
Comment (layershifter) [inline packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]:
```suggestion
  const handleOnMotionFinish = useEventCallback(
```

nit to be consistent with naming in other files

File: packages/react-components/react-utilities/src/hooks/useAnimationFrame.ts [hunk 11]
```diff
@@ -17,12 +17,12 @@ const cancelAnimationFrameNoop = (handle: number) => handle;
  *
  * @returns A pair of [requestAnimationFrame, cancelAnimationFrame] that are stable between renders.
  */
-export function useAnimationFrame(): readonly [(fn: () => void, delay?: number) => number, () => void] {
+export function useAnimationFrame(): readonly [(fn: FrameRequestCallback) => number, () => void] {
```
Comment (layershifter) [inline packages/react-components/react-utilities/src/hooks/useAnimationFrame.ts:20]: Can you please create a separate PR for this?

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts (hunk 8 above)
Comment (layershifter) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: image

I'm referring to this pattern. The current implementation expects properties like `visible`, `delay`, etc. to be explicitly defined in **props** 💥 As a result, `acceptsVisibleProp(<TestComponent />)` will return `false` even if the component accepts `visible`.

To clarify, manipulating children has never been a strongly recommended pattern in React. That's why it's considered legacy, and as the React docs state, ["_Using Children is uncommon and can lead to fragile code_"](https://react.dev/reference/react/Children) 🐱 
For this reason, other APIs in Fluent v9 try to avoid using these patterns as much as possible.

However, when dealing with grouped motions (_and keeping items mounted_), things get tricky. This is why `PresenceGroup` uses `React.Children`. It’s an unavoidable evil in this particular case 😭 

Back to custom components: I agree that it's fine to accept our primitives as they are. However, for custom components, I believe it’s crucial to have an **explicit** API (_that's why I suggest using a factory_) to avoid confusion and potential issues when devs will modify examples from our docs and accidentally nothing works. [Explicit is better than implicit](https://peps.python.org/pep-0020/)

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx (hunk 10 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/useStaggerItemsVisibility.tsx]: Done.

File: packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts (hunk 8 above)
Comment (robertpenner) [reply packages/react-components/react-motion-components-preview/library/src/choreography/Stagger/utils/motionComponentDetection.ts]: Done.
For this PR, I removed `hasExplicitProps` and the auto-detection will just focus on our motion components.
The factory can be added in a future PR.

File: packages/react-components/react-utilities/src/hooks/useAnimationFrame.ts (hunk 11 above)
Comment (robertpenner) [reply packages/react-components/react-utilities/src/hooks/useAnimationFrame.ts:20]: Done.
New PR for `useAnimationFrame` types is #35466.

**Review by layershifter — APPROVED** [review]
